framework/test-vectors/test-vector-enumeration.md

```diff
@@ -0,0 +1,176 @@
+
+## Expected results
+
+An expected result is a categorical value
+and an optional error description.
+
+An expected result categorical value is one of:
+* Positive (successful test scenario)
+    * `"positive-keyring"`
+* Negative encrypt (failure on encrypt)
+    * `"negative-encrypt-keyring"`
+* Negative decrypt (failure on decrypt)
+    * `"negative-decrypt-keyring"`
+
+The error description for negative scenarios
+describes the reason for the scenario's failure.
+
+### Determining expected results
+
+In contrast to [input configuration enumeration](#enumerating-input-configurations),
+which is a constructive multiplicative process,
+the expected result of a test
+is deterministic from the input configuration.
+
+The expected result of an [input configuration](#input-configuration) can be determined
+by evaluating all relevant [evaluation rules](#expected-result-evaluation-rules) for that input configuration.
+If no evaluation rules fail, the test scenario result should be `"positive-keyring"`.
+If one evaluation rule fails, the test scenario result should be the result of that evaluation rule.
+If multiple evaluation rules fail, the test scenario should not be written. See [below](#multiple-errors-todo).
+
+#### Expected result evaluation rules
+
+An evaluation rule reads the input configuration [input configuration](#input-configuration),
+and determines whether the configuration is valid
+under the conditions the rule evaluates.
+
+For example,
+the [required encryption context CMM component](complete-vectors/required-encryption-context-cmm.md#required-encryption-context-cmm-failures-on-encrypt)
+defines the conditions for encryption to fail.
+This can be interpreted as an evaluation rule,
+and the test manifest generator should implement this
+when generating test vectors using the required encryption context CMM
+to determine when these vectors should fail.
+
+These rules can be relatively informal in the spec,
+but ideally should be written programmatically,
+so a developer can write implement these rules in test manifest generation code.
+
+Evaluation rules are a useful abstraction to determine a given test scenario's expected result. A list of independent rules is flexible and maintainable. A list of rules that maintain the same interface (reads input configuration; outputs success/fail) makes evaluating these rules programmable.
+
+### Multiple errors (TODO)
+
+An input configuration SHOULD only be deterministic of a single error.
+If a single input configuration can result in multiple errors,
+it SHOULD NOT be written.
+For example,
+if a test scenario specifies that `decrypt` provides incorrect `reproducedEncryptionContext` to a requried encryption context CMM
+but also specifies a key that cannot decrypt the encrypted value,
+that scenario should not be specified.
+
+(Is the above accurate?
+Do we need some method to determine precedence of errors?
```

Comment:
IMO, not in test vectors. This feels more like something we'd use per-language integ/functional tests for, since those are already more specific to individual runtime releases as opposed to the more general test-vectors.

Reply:
Yep, agreed.
Ryan and I chatted and decided this might be nice at some point in the very distant future, but is not worth investment right now.
In the meantime, we'll say that any "error description" is just a best-effort attempt to describe the error and should not be trusted.
(but if the error type is both negative-encrypt and negative-decrypt, write negative-encrypt.)
In the future, we might add some new error description field that is "trusted", but not right now.